src/globus_cli/login_manager.py

```diff
@@ -0,0 +1,93 @@
+import functools
+
+import click
+
+from .tokenstore import token_storage_adapter
+
+
+class LoginManager:
+    def __init__(self):
+        self._token_storage = token_storage_adapter()
+
+    def has_login(self, resource_server: str):
+        """
+        Determines if the user has a refresh token for the given
+        resource server
+        """
+        tokens = self._token_storage.get_token_data(resource_server)
+        if tokens is None or "refresh_token" not in tokens:
+            return False
+
+        return True
+
+
+def requires_login(*args: str, pass_manager: bool = False):
+    """
+    Command decorator for specifying a resource server that the user must have
+    tokens for in order to run the command.
+
+    Simple usage for commands that have static resource needs: simply list all
+    needed resource servers as args:
+
+    @requries_login("auth.globus.org")
+
+    @requires_login("auth.globus.org", "transfer.api.globus.org")
+
+    Usage for commands which have dynamic resource servers depending
+    on the arguments passed to the command (e.g. commands for the GCS API)
+
+    @requies_login(pass_manager=True)
+    def command(login_manager, endpoint_id)
+
+        login_manager.<do the thing>(endpoint_id)
+
+    """
+    resource_servers = args
+
+    def inner(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            manager = LoginManager()
+
+            # determine the set of resource servers missing logins
+            missing_servers = set()
+            for server_name in resource_servers:
+                if not manager.has_login(server_name):
+                    missing_servers.add(server_name)
+
+            # if we are missing logins, assemble error text
+            # text is slightly different for 1, 2, or 3+ missing servers
+            if missing_servers:
+
+                if len(missing_servers) == 1:
+                    plural_string = ""
+                    server_string = missing_servers.pop()
```

Comment:
I might factor the formatting-focused bits out into a utils module.

Reply:
That's fair, it got more complicated than I intended. Not sure how much we even care about listing all the missing servers. Might just be better to say the user needs to log in again until we change the login command to accept arbitrary resource servers.

Reply:
I think we've done something like this for output in the CLI before, FWIW.
If you're making a utility, it might be worth taking a glance around to see if we can have a helper which does this.

I think `pluralize(value: List[str]) -> Tuple[bool, str]` which returns `(is_plural, list_formatted_string)` might be good. Where to apply that `"s"` in the general case could get tricky.

On a related note, I think it's fine to print `Missing login(s) for ...` to dodge that part of the issue.